crates/bevy_core_pipeline/src/bloom/mod.rs

```diff
@@ -360,199 +445,31 @@ impl FromWorld for BloomPipelines {
-                    blend: None,
-                    write_mask: ColorWrites::ALL,
-                })],
-            }),
-            primitive: PrimitiveState::default(),
-            depth_stencil: None,
-            multisample: MultisampleState::default(),
-        });
-
-        let upsampling_final_pipeline =
-            pipeline_cache.queue_render_pipeline(RenderPipelineDescriptor {
-                label: Some("bloom_upsampling_final_pipeline".into()),
-                layout: Some(vec![downsampling_bind_group_layout.clone()]),
-                vertex: fullscreen_shader_vertex_state(),
-                fragment: Some(FragmentState {
-                    shader: BLOOM_SHADER_HANDLE.typed::<Shader>(),
-                    shader_defs: vec![],
-                    entry_point: "upsample_final".into(),
-                    targets: vec![Some(ColorTargetState {
-                        format: ViewTarget::TEXTURE_FORMAT_HDR,
-                        blend: Some(BlendState {
-                            color: BlendComponent {
-                                src_factor: BlendFactor::One,
-                                dst_factor: BlendFactor::One,
-                                operation: BlendOperation::Add,
-                            },
-                            alpha: BlendComponent::REPLACE,
-                        }),
-                        write_mask: ColorWrites::ALL,
-                    })],
-                }),
-                primitive: PrimitiveState::default(),
-                depth_stencil: None,
-                multisample: MultisampleState::default(),
-            });
+        BloomPipelines { sampler }
+    }
+}
 
-        BloomPipelines {
-            downsampling_prefilter_pipeline,
-            downsampling_pipeline,
-            upsampling_pipeline,
-            upsampling_final_pipeline,
-            sampler,
-            downsampling_bind_group_layout,
-            upsampling_bind_group_layout,
+fn extract_bloom_settings(
+    mut commands: Commands,
+    cameras: Extract<Query<(Entity, &Camera, &BloomSettings), With<Camera>>>,
+) {
+    for (entity, camera, bloom_settings) in &cameras {
+        if camera.is_active {
+            commands.get_or_spawn(entity).insert(bloom_settings.clone());
         }
     }
 }
 
 #[derive(Component)]
-struct BloomTextures {
-    texture_a: CachedTexture,
-    texture_b: CachedTexture,
+struct BloomTexture {
+    // First mip is half the screen resolution, successive mips are half the previous
+    texture: CachedTexture,
     mip_count: u32,
 }
 
-impl BloomTextures {
-    fn texture_view(texture: &CachedTexture, base_mip_level: u32) -> TextureView {
-        texture.texture.create_view(&TextureViewDescriptor {
+impl BloomTexture {
+    fn view(&self, base_mip_level: u32) -> TextureView {
+        self.texture.texture.create_view(&TextureViewDescriptor {
             base_mip_level,
             mip_level_count: Some(unsafe { NonZeroU32::new_unchecked(1) }),
```

Comment:
There must have been a reason this was not done without unsafe in the first place, right? I'll commit this but if that breaks something it's on you 😛

Reply:
I wrote this code originally - the reason was because I didn't know NonZeroU32 had a new() method to begin with :P. 

It's not going to break anything either way - the only way it could is by trying to create a NonZeroU32 set to 0. This code is slightly simpler, removes a dumb usage of "unsafe" that's pretty unnecessary, and the compiler should optimize the check (!= 0) out anyways.